diff --git a/apps/konsole/src/TerminalDisplay.cpp b/apps/konsole/src/TerminalDisplay.cpp
index 1f660e8..aa47d79 100644
--- a/apps/konsole/src/TerminalDisplay.cpp
+++ b/apps/konsole/src/TerminalDisplay.cpp
@@ -1592,6 +1592,7 @@ void TerminalDisplay::updateImageSize()
   if ( _resizing )
   {
       showResizeNotification();
+      processFilters(); // Obtain new hotspot indicies
     emit changedContentSizeSignal(_contentHeight, _contentWidth); // expose resizeEvent
   }
   
