diff -up kdebase-4.3.4/apps/konsole/src/MainWindow.cpp.me kdebase-4.3.4/apps/konsole/src/MainWindow.cpp
--- kdebase-4.3.4/apps/konsole/src/MainWindow.cpp.me	2012-07-04 22:02:44.033364676 +0200
+++ kdebase-4.3.4/apps/konsole/src/MainWindow.cpp	2012-07-04 22:05:40.841389410 +0200
@@ -252,7 +252,6 @@ void MainWindow::setupActions()
     _toggleMenuBarAction = new KToggleAction(this);
     _toggleMenuBarAction->setText( i18n("Show Menu Bar") );
     _toggleMenuBarAction->setIcon( KIcon("show-menu") );
-    _toggleMenuBarAction->setChecked( !menuBar()->isHidden() );
     connect( _toggleMenuBarAction , SIGNAL(toggled(bool)) , menuBar() , SLOT(setVisible(bool)) );
     collection->addAction("show-menubar",_toggleMenuBarAction);
 
@@ -476,5 +475,17 @@ void MainWindow::configureNotifications(
     KNotifyConfigWidget::configure( this );
 }
 
+void MainWindow::showEvent(QShowEvent *event)
+{
+    // This code from Konqueror.
+    // We need to check if our toolbars are shown/hidden here, and set
+    // our menu items accordingly. We can't do it in the constructor because
+    // view profiles store toolbar info, and that info is read after
+    // construct time.
+    _toggleMenuBarAction->setChecked( !menuBar()->isHidden() );
+    // Call parent method
+    KXmlGuiWindow::showEvent(event);
+}
+
 #include "MainWindow.moc"
 
diff -up kdebase-4.3.4/apps/konsole/src/MainWindow.h.me kdebase-4.3.4/apps/konsole/src/MainWindow.h
--- kdebase-4.3.4/apps/konsole/src/MainWindow.h.me	2012-07-04 21:58:57.257361284 +0200
+++ kdebase-4.3.4/apps/konsole/src/MainWindow.h	2012-07-04 22:05:44.922387391 +0200
@@ -141,6 +141,9 @@ class MainWindow : public KXmlGuiWindow
         void closeActiveSessionRequest();
 
     protected:
+        // Reimplemented for internal reasons.
+        virtual void showEvent(QShowEvent *event);
+
         // reimplemented from KMainWindow
         virtual bool queryClose();
         virtual void saveProperties(KConfigGroup& group);
