diff --git a/apps/konsole/src/MainWindow.cpp b/apps/konsole/src/MainWindow.cpp
index ea53c26..de36e23 100644
--- a/apps/konsole/src/MainWindow.cpp
+++ b/apps/konsole/src/MainWindow.cpp
@@ -41,6 +41,7 @@
 #include <KToolInvocation>
 #include <KStandardAction>
 #include <KStandardGuiItem>
+#include <KWindowSystem>
 #include <KXMLGUIFactory>
 #include <KNotifyConfigWidget>
 
@@ -350,6 +351,12 @@ bool MainWindow::queryClose()
         _viewManager->viewProperties().count() < 2)
         return true;
 
+    // make sure the window is shown on current desktop and is not minimized
+    KWindowSystem::setOnDesktop(winId(), KWindowSystem::currentDesktop());
+    if ( isMinimized() ) {
+        KWindowSystem::unminimizeWindow(winId(), true);
+    }
+
     int result = KMessageBox::warningYesNoCancel(this,
                 i18n("You have multiple tabs in this window, "
                      "are you sure you want to quit?"),
