diff --git a/apps/konsole/src/TerminalDisplay.cpp b/apps/konsole/src/TerminalDisplay.cpp
index 1f660e8..cd67b3c 100644
--- a/apps/konsole/src/TerminalDisplay.cpp
+++ b/apps/konsole/src/TerminalDisplay.cpp
@@ -319,7 +319,7 @@ TerminalDisplay::TerminalDisplay(QWidget *parent)
 ,_resizeTimer(0)
 ,_flowControlWarningEnabled(false)
 ,_outputSuspendedLabel(0)
-,_lineSpacing(0)
+,_lineSpacing(1)
 ,_colorsInverted(false)
 ,_blendColor(qRgba(0,0,0,0xff))
 ,_filterChain(new TerminalImageFilterChain())
@@ -710,7 +710,7 @@ void TerminalDisplay::drawCharacters(QPainter& painter,
         if (_bidiEnabled)
             painter.drawText(rect,0,text);
         else
-            painter.drawText(rect,0,LTR_OVERRIDE_CHAR+text);
+            painter.drawText(rect,Qt::AlignBottom,LTR_OVERRIDE_CHAR+text);
     }
 }
 
