diff --git a/apps/plasma/applets/folderview/folderview.cpp b/apps/plasma/applets/folderview/folderview.cpp
index 821d19c..9efbecc 100644
--- a/apps/plasma/applets/folderview/folderview.cpp
+++ b/apps/plasma/applets/folderview/folderview.cpp
@@ -47,6 +47,7 @@
 #include <KStandardShortcut>
 #include <KStringHandler>
 #include <KTemporaryFile>
+#include <ktoolinvocation.h>
 
 #include <kio/copyjob.h>
 #include <kio/fileundomanager.h>
@@ -1324,8 +1325,13 @@ void FolderView::createActions()
     logoutAction->setIcon(KIcon("system-shutdown"));
     connect(logoutAction, SIGNAL(triggered(bool)), this, SLOT(logout()));
 
+    QAction *runKonsoleAction = new QAction(i18n("Konsole"), this);
+    runKonsoleAction->setIcon(KIcon("utilities-terminal"));
+    connect(runKonsoleAction, SIGNAL(triggered(bool)), this, SLOT(runKonsole()));
+
     m_actionCollection.addAction("add panel", addPanelAction);
     m_actionCollection.addAction("run command", runCommandAction);
+    m_actionCollection.addAction("run konsole", runKonsoleAction);
     m_actionCollection.addAction("lock screen", lockScreenAction);
     m_actionCollection.addAction("logout", logoutAction);
 
@@ -1448,18 +1454,19 @@ QList<QAction*> FolderView::contextualActions()
     if (isContainment()) {
         actions.append(action("add widgets"));
         actions.append(m_actionCollection.action("add panel"));
+        actions.append(m_actionCollection.action("run konsole"));
         if (KAuthorized::authorizeKAction("run_command")) {
             actions.append(m_actionCollection.action("run command"));
         }
         if (screen() == -1) {
             actions.append(action("remove"));
-        }   
+        }
         actions.append(action("lock widgets"));
 
         QAction *separator = new QAction(this);
         separator->setSeparator(true);
         actions.append(separator);
- 
+
         if (KAuthorized::authorizeKAction("lock_screen")) {
             actions.append(m_actionCollection.action("lock screen"));
         }
@@ -1554,6 +1561,11 @@ void FolderView::runCommand()
     krunner.call("display");
 }
 
+void FolderView::runKonsole()
+{
+    KToolInvocation::invokeTerminal(QString(), QDir::homePath());
+}
+
 void FolderView::lockScreen()
 {
     if (!KAuthorized::authorizeKAction("lock_screen")) {
diff --git a/apps/plasma/applets/folderview/folderview.h b/apps/plasma/applets/folderview/folderview.h
index e012f4b..a389ec4 100644
--- a/apps/plasma/applets/folderview/folderview.h
+++ b/apps/plasma/applets/folderview/folderview.h
@@ -132,6 +132,7 @@ protected slots:
     void addPanel();
     void addPanel(const QString &plugin);
     void runCommand();
+    void runKonsole();
     void lockScreen();
     void logout();
 

